rapid7/metasploit-framework
# Lucee Scheduled Job RCE

Lucee is an Open Source ColdFusion server/engine intended for rapid web development. Many implementations of
ColdFusion files support dynamic input and server side code execution.
In the case of this module, Lucees implementation supports the use of `cfexecute` and `cfscript` tags in `.cfm` files.

In addition to these features, Lucee provides a scheduled job feature. This feature will accept an
external `url` argument and query that page on execution. If logging is enabled, it is possible to
query a remote ColdFusion document, log it in the web root, and access it to execute its code,
subsequently achieving arbitrary server side code execution. The series of requests to achieve this is as follows.

1. Authenticate as the administrator to the web admin panel
2. Create a scheduled job that includes a URL to the remote ColdFusion document
3. Update the scheduled job to turn on logging and ensure that the remote document is logged to the web root
4. Execute the scheduled job. The Lucee server will now reach out to and download the ColdFusion document from the attackers server
5. Access the document at the web root of the server, thus executing the payload.

The basic format for the remote ColdFusion document is as follows.

```html
<cfscript>
            cfexecute(name="powershell.exe", arguments="-c whoami",timeout=5);
</cfscript>
```
## Verification

List the steps needed to make sure this thing works

Exact steps may vary depending on the choice of payload. However, a general outline for receiving a reverse shell (independent of platform) is below.

1. Download and install Lucee from the vendors website. This can be done on either a Windows or Unix host.
   No additional setup is needed beyond the initial installation walkthrough
2. Start MSF Console
3. Do: `use multi/http/lucee_scheduled_job_rce`
4. Choose a target that reflects the target system
	- `use X` (0 for Windows, 1 for Linux)
5. Select payload. This functions with command execution payloads and supports reverse shells and generic commands.
6. Select the desired payload and complete its requirement. `CMD`, `LHOST`, `LPORT`, etc.
7. Select the appropriate `RHOST`. `PASSWORD`, and (if necessary), the `TARGETURI`
8. Execute the payload. You should either receive a shell or see the output of your command.

For the sake of demonstration, I've included a quick demo video in an email titled _Lucee Scheduled Jobs RCE_

Setting up a testing environment does not require any additional setup or configuration beyond the initial installation walkthrough. Downloads for both Windows and Unix can be found [here](https://download.lucee.org/)

Thank you, and I'm looking forward to hearing back!
